charts: add allowPrivilegeEscalation: true to containerSecurityContext to nodeplugin daemonset

losil opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

When running the kubernetes cluster with one single privileged PodSecurityPolicy which is allowing everything the nodeplugin daemonset can fail to start. To be precise the problem is the `defaultAllowPrivilegeEscalation: false` configuration in the PSP. Containers of the nodeplugin daemonset won't start when they have `privileged: true` but no `allowPrivilegeEscalation` in their container securityContext.

Kubernetes will not schedule if this mismatch exists `cannot set allowPrivilegeEscalation to false and privileged to true`:
```
[Warning  FailedCreate  7s (x3 over 14s)  daemonset-controller  (combined from similar events): Error creating: Pod "ceph-csi-nodeplugin-rbd-7qm9f" is invalid: [spec.containers[0].securityContext: Invalid value: core.SecurityContext{Capabilities:(*core.Capabilities)(nil), Privileged:(*bool)(0xc04f579a0c), SELinuxOptions:(*core.SELinuxOptions)(nil), WindowsOptions:(*core.WindowsSecurityContextOptions)(nil), RunAsUser:(*int64)(nil), RunAsGroup:(*int64)(nil), RunAsNonRoot:(*bool)(nil), ReadOnlyRootFilesystem:(*bool)(nil), AllowPrivilegeEscalation:(*bool)(0xc04d20ce50), ProcMount:(*core.ProcMountType)(nil), SeccompProfile:(*core.SeccompProfile)(nil)}: cannot set `allowPrivilegeEscalation` to false and `privileged` to true, spec.containers[2].securityContext: Invalid value: core.SecurityContext{Capabilities:(*core.Capabilities)(nil), Privileged:(*bool)(0xc04f579a0f), SELinuxOptions:(*core.SELinuxOptions)(nil), WindowsOptions:(*core.WindowsSecurityContextOptions)(nil), RunAsUser:(*int64)(nil), RunAsGroup:(*int64)(nil), RunAsNonRoot:(*bool)(nil), ReadOnlyRootFilesystem:(*bool)(nil), AllowPrivilegeEscalation:(*bool)(0xc04d20ce50), ProcMount:(*core.ProcMountType)(nil), SeccompProfile:(*core.SeccompProfile)(nil)}: cannot set `allowPrivilegeEscalation` to false and `privileged` to true]]()
``` 

The default PodSecurityPolicy for every workload in the k8s cluster is the following:
```yaml
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  annotations:
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
  name: system-unrestricted-psp
spec:
  allowPrivilegeEscalation: true
  allowedCapabilities:
  - '*'
  defaultAllowPrivilegeEscalation: false
  fsGroup:
    rule: RunAsAny
  hostIPC: true
  hostNetwork: true
  hostPID: true
  hostPorts:
  - max: 65535
    min: 0
  privileged: true
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  volumes:
  - '*'
```

The daemonsets run when the `allowPrivilegeEscalation` parameter is added to the container securityContext config. In detail this is needed for the containers:
* driver-registrar
* liveness-prometheus

Updating the helm chart templates for `ceph-csi-rbd` and `ceph-csi-cephfs`solve the problem when using `defaultAllowPrivilegeEscalation: true` in global privileged PodSecurityPolicy. For example:

```yaml
{{- if .Values.nodeplugin.httpMetrics.enabled }}
        - name: liveness-prometheus
          securityContext:
            privileged: true
            allowPrivilegeEscalation: true
          image: "{{ .Values.nodeplugin.plugin.image.repository }}:{{ .Values.nodeplugin.plugin.image.tag }}"
``` 


## Is there anything that requires special attention ##

Do you have any questions? No

Is the change backward compatible? Yes

Are there concerns around backward compatibility? No

Provide any external context for the change, if any.

For example:

https://kubernetes.io/docs/tasks/configure-pod-container/security-context/


## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

None

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

None

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

losil committed:
helm: allowPrivilegeEscalation: true in containerSecurityContext

When running the kubernetes cluster with one single privileged
PodSecurityPolicy which is allowing everything the nodeplugin
daemonset can fail to start. To be precise the problem is the
defaultAllowPrivilegeEscalation: false configuration in the PSP.
 Containers of the nodeplugin daemonset won't start when they
have privileged: true but no allowPrivilegeEscalation in their
container securityContext.

Kubernetes will not schedule if this mismatch exists cannot set
allowPrivilegeEscalation to false and privileged to true

Signed-off-by: Silvan Loser <silvan.loser@hotmail.ch>
Signed-off-by: Silvan Loser <33911078+losil@users.noreply.github.com>

yati1998 commented:
@losil please check the DCO failure. We have specific guidelines for the commit messages.

humblec commented:
@losil as you know  PSP  ( PodSecurityPolicy)  is getting deprecated in upstream.  Regardless we could have this in chart :+1:

losil commented:
> @losil as you know PSP ( PodSecurityPolicy) is getting deprecated in upstream. Regardless we could have this in chart 👍

@humblec of course I am aware of the PSPs deprecation and honestly look forward to getting rid of them.
But since we are still running our clusters with k8s `1.21.10` we still rely on the PSPs.

humblec commented:
Yeah , please correct the commit lint error @losil .. it should be good to go..

Madhu-1 commented:
@losil Please check https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages standard followed for commit message.

it should be something like below

```
helm: allowPrivilegeEscalation: true in containerSecurityContext

When running the kubernetes cluster with one single privileged 
PodSecurityPolicy which is allowing everything the nodeplugin 
daemonset can fail to start. To be precise the problem is the 
defaultAllowPrivilegeEscalation: false configuration in the PSP.
 Containers of the nodeplugin daemonset won't start when they 
have privileged: true but no allowPrivilegeEscalation in their c
ontainer securityContext.

Kubernetes will not schedule if this mismatch exists cannot set 
allowPrivilegeEscalation to false and privileged to true:

Signed-off-by: Silvan Loser <silvan.loser@hotmail.ch>
```

losil commented:
> @losil Please check https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages standard followed for commit message.
> 
> it should be something like below
> 
> ```
> helm: allowPrivilegeEscalation: true in containerSecurityContext
> 
> When running the kubernetes cluster with one single privileged 
> PodSecurityPolicy which is allowing everything the nodeplugin 
> daemonset can fail to start. To be precise the problem is the 
> defaultAllowPrivilegeEscalation: false configuration in the PSP.
>  Containers of the nodeplugin daemonset won't start when they 
> have privileged: true but no allowPrivilegeEscalation in their c
> ontainer securityContext.
> 
> Kubernetes will not schedule if this mismatch exists cannot set 
> allowPrivilegeEscalation to false and privileged to true:
> 
> Signed-off-by: Silvan Loser <silvan.loser@hotmail.ch>
> ```

Thanks for the hint!

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio refresh

Madhu-1 reviewed (changes requested):
@losil Sorry for the late review on this one.

Same changes are required in the below files

deploy/cephfs/kubernetes/csi-cephfsplugin.yaml
deploy/rbd/kubernetes/csi-rbdplugin.yaml

losil committed:
deploy: allowPrivilegeEscalation: true in containerSecurityContext

When running the kubernetes cluster with one single privileged
PodSecurityPolicy which is allowing everything the nodeplugin
daemonset can fail to start. To be precise the problem is the
defaultAllowPrivilegeEscalation: false configuration in the PSP.
 Containers of the nodeplugin daemonset won't start when they
have privileged: true but no allowPrivilegeEscalation in their
container securityContext.

Kubernetes will not schedule if this mismatch exists cannot set
allowPrivilegeEscalation to false and privileged to true:

Signed-off-by: Silvan Loser <silvan.loser@hotmail.ch>
Signed-off-by: Silvan Loser <33911078+losil@users.noreply.github.com>

Madhu-1 commented:
@losil can you please change `Signed-off-by: Silvan Loser <33911078+losil@users.noreply.github.com>` to `Signed-off-by: Silvan Loser <silvan.loser@hotmail.ch>` in your 2nd commit?

humblec commented:
@losil can you please revisit this PR mainly for correcting `DCO`?

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio rebase

Madhu-1 reviewed (commented):
@losil changes LGTM, can you please rebase the PR, PR is having 3 commits, it should have only 2

Madhu-1 commented:
/retest ci/centos/mini-e2e/k8s-1.23

Madhu-1 commented:
>This branch cannot be rebased due to conflicts

This PR is having merge conflicts needs to be addressed with manual rebase

Madhu-1 reviewed (changes requested):
https://github.com/ceph/ceph-csi/pull/2993#issuecomment-1106223019

humblec commented:
> > This branch cannot be rebased due to conflicts
> 
> This PR is having merge conflicts needs to be addressed with manual rebase

@Madhu-1 where do you see the conflict?

Madhu-1 commented:
> > > This branch cannot be rebased due to conflicts
> > 
> > 
> > This PR is having merge conflicts needs to be addressed with manual rebase
> 
> @Madhu-1 where do you see the conflict?

After PR approval it is showing conflict and cannot be rebased.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
> > rebase
> 
> ####  Base branch update has failed

@humblec please see here.

humblec commented:
I am not able to find @losil in ceph csi slack too.. Lets go ahead without this PR then.. :+1:

losil commented:
Hey @Madhu-1 

Hope it's fine now.. Sorry for the delay... I'm on vacation.

Cheers

Madhu-1 commented:
@losil no worries, Thanks for updating the PR. Will add backport label once 3.6.1 is released.

Madhu-1 commented:
@mergifyio refresh
